src/utils/java_string_formatter.hpp

```diff
@@ -0,0 +1,101 @@
+// Copyright 2023 Memgraph Ltd.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#pragma once
+
+#include <string>
+
+#include "utils/exceptions.hpp"
+#include "utils/pmr/string.hpp"
+#include "utils/pmr/vector.hpp"
+
+namespace memgraph::utils {
+
+class JStringFormatException final : public BasicException {
+ public:
+  explicit JStringFormatException(const std::string &what) noexcept : BasicException(what) {}
+
+  template <class... Args>
+  explicit JStringFormatException(fmt::format_string<Args...> fmt, Args &&...args) noexcept
+      : JStringFormatException(fmt::format(fmt, std::forward<Args>(args)...)) {}
+};
+
+template <typename T>
+concept TTypedValueLike = requires(T t) {
+  { t.ValueInt() } -> std::convertible_to<int>;
+  { t.ValueDouble() } -> std::convertible_to<double>;
+  { t.ValueString() } -> std::convertible_to<pmr::string>;
+};
+
+template <typename TString, typename TTypedValueLike>
+class JStringFormatter final {
+ public:
+  [[nodiscard]] TString FormatString(TString str, const pmr::vector<TTypedValueLike> &format_args) const {
```

Comment:
Can you catch string here by reference?

Reply:
We could, but we do a bunch of in-place modifications on the string so a copy would be needed either way. We could catch by non-const lvalue& but that would change the semantics of the function. This function should only be used to do Java like string formatting, we will do it very rearly. I would like to keep the function signature more ergonomic rather then one-less-string-copy faster.